### PR TITLE
Add GroupItemStateChangeEvent #596

### PR DESCRIPTION
**Pull request by chriskn**
Assertions:
If a member of a group is changed then notify the observers of the member as well as the observers of the group about the changes.
If a member is added or removed from a group then notify the observers of the group about the added or removed member.
Observers of the group are informed about the name of the changed, added or removed member.
If one member is added or removed then one group change is fired.
If one member changes two changes are fired: one change for the member itself and one for the group.

Implementation:
GroupItems hold a State of type GroupType which contains the name of the added, removed or changed member and the type of the change as enumeration constant of type GroupChange (ADDED, REMOVED, CHANGED)
If a member is added, removed or changed the group notifies its observers about the changes. Therefore the method notifyObservers of class GenericItem is used. In order to use this method the visibility was changed from private to protected.
Extended the GroupItemTest to test the notification of observers.

Bug: Add GroupItemStateChangeEvent #596
Signed-off-by: Christoph Knauf  <christoph.knauf@itemis.de>

### CODE REVIEW

**Comment by maggu2810** [comment]
Wow! I love such detailed commit messages. Thank you

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java [hunk 1]
```diff
@@ -244,6 +251,8 @@ public String toString() {
     @Override
     public void stateChanged(Item item, State oldState, State newState) {
         setState(function.calculate(members));
+        setGroupState(new GroupType(item.getName(),GroupChange.CHANGED));
+
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: IMHO we should remove the empty line in front of a closing bracket.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java [hunk 2]
```diff
@@ -252,6 +261,27 @@ public void stateChanged(Item item, State oldState, State newState) {
     @Override
     public void stateUpdated(Item item, State state) {
         setState(function.calculate(members));
+        setGroupState(new GroupType(item.getName(),GroupChange.CHANGED));
+
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: IMHO we should remove the empty line in front of a closing bracket.

**Comment by maggu2810** [comment]
I don't know if we should use the state to hold that information.
It seems like a workaround if the state of a group item consists of the last action.

Not to be used here, but to explain what I mean:
That is (IMHO) the difference between states and commands. A add, remove, change command does make sense and a state that contains e.g. the group members. But a state "removed item1"... Don't know... It just saves the last action that is done.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java (hunk 2 above)
Comment (chriskn) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: Removed empty lines

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java (hunk 1 above)
Comment (chriskn) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: Removed empty lines

**Comment by maggu2810** [comment]
IMHO there is a analogy to the DimmerItem. The state does not contain the information "increased" to signal that the last action has been "increase" [command]. But the state is normally a percent type and that one should be changed on an increase.

**Comment by chriskn** [comment]
So would that mean for a groupe that each state should store the members of the group? 
Then you have to compare the number of members in newState and oldState in order to find out if it was an add or a remove event. For change events this would not work.

**Comment by maggu2810** [comment]
Yes, that would not work.
I see the point that we need an event if a item is added to / removed from a group or is changed by the group item observer.
But I don't think that it is correct to implement the state of an group that way.

Perhaps you already discussed that with @kaikreuzer and my understanding of a "state" is not correct.
But modeling a that that it only contains the last action and not a state is confusing.
A state of group that consists of item_a item_b item_c could be "removed item_d" or "added item_c". That is confusing...

**Comment by kaikreuzer** [comment]
FTR: I just discussed with @chriskn that the idea of #596 was not about events if members are added or removed, but about events of the (already existing) state of the group item, which is calculated from the state of its members. He will update the PR accordingly.

**Comment by chriskn** [comment]
I fixed the test and leave the comment so you can comprehend why the old state is stored in the group item

File: .gitignore [hunk 3]
```diff
@@ -15,3 +15,4 @@ distribution/smarthome/runtime/etc/mapdb/storage.mapdb.t
 bundles/storage/org.eclipse.smarthome.storage.mapdb.test/userdata/mapdb/storage.mapdb
 bundles/storage/org.eclipse.smarthome.storage.mapdb.test/userdata/mapdb/storage.mapdb.p
 bundles/storage/org.eclipse.smarthome.storage.mapdb.test/userdata/mapdb/storage.mapdb.t
+.project
```
Comment (maggu2810) [inline .gitignore]: I don't think we would like to ignore Eclipse project files.

File: targetplatform/smarthome.target [hunk 4]
```diff
@@ -130,7 +131,7 @@
 </location>
 <location includeAllPlatforms="false" includeConfigurePhase="false" includeMode="planner" includeSource="true" type="InstallableUnit">
 <unit id="javax.jmdns.feature.feature.group" version="3.4.2"/>
-<repository location="http://www.openhab.org/jmdns/update-site/3.4.2/"/>
+<repository location="http://www.jmdns.org/update-site/3.4.2/"/>
```
Comment (maggu2810) [inline targetplatform/smarthome.target]: This should not be part of this PR.

**Comment by maggu2810** [comment]
> was not about events if members are added or removed, but about events of the (already existing) state of the group item

The PR still contains GroupType.GroupChange.ADDED, ...
I believe I didn't get the point what this PR is about.

**Comment by chriskn** [comment]
> The PR still contains GroupType.GroupChange.ADDED, ...

Sorry, I forgot to delete it

File: targetplatform/smarthome.target [hunk 5]
```diff
@@ -55,6 +55,7 @@
 <unit id="org.codehaus.jackson.mapper.source" version="1.6.0.v20101005-0925"/>
 <unit id="org.hamcrest.core" version="1.3.0.v201303031735"/>
 <unit id="org.hamcrest.core.source" version="1.3.0.v201303031735"/>
+<unit id="org.json" version="1.0.0.v201011060100"/>
```
Comment (maggu2810) [inline targetplatform/smarthome.target]: Why have you changed the target platform? Cannot found a new import that contains "json" in your diff.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java [hunk 6]
```diff
@@ -251,7 +255,20 @@ public void stateChanged(Item item, State oldState, State newState) {
      */
     @Override
     public void stateUpdated(Item item, State state) {
+        oldGroupState = this.state; 
         setState(function.calculate(members));
+        sendGroupStateUpdatedEvent(item.getName(), this.state);
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: Shouldn't we check if this.oldGroupState and this.state differs and send a ChangedEvent if it differs (instead of always an update).
As long as the group state is calculated by the members it could differ (perhaps relevant to the used function, e.g. state change counter).

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/events/ItemEventFactory.java [hunk 7]
```diff
@@ -70,10 +74,32 @@ protected Event createEventByType(String eventType, String topic, String payload
             event = createUpdatedEvent(topic, payload);
         } else if (eventType.equals(ItemRemovedEvent.TYPE)) {
             event = createRemovedEvent(topic, payload);
+        } else if (eventType.equals(GroupItemStateChangedEvent.TYPE)){
+            event = createGroupStateChangedEvent(topic, payload); 
+        }else if (eventType.equals(GroupItemStateEvent.TYPE)){
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/events/ItemEventFactory.java]: missing space after } and in front of {

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/events/ItemEventFactory.java [hunk 8]
```diff
@@ -70,10 +74,32 @@ protected Event createEventByType(String eventType, String topic, String payload
             event = createUpdatedEvent(topic, payload);
         } else if (eventType.equals(ItemRemovedEvent.TYPE)) {
             event = createRemovedEvent(topic, payload);
+        } else if (eventType.equals(GroupItemStateChangedEvent.TYPE)){
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/events/ItemEventFactory.java]: missing space in front of {

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/events/ItemEventFactory.java [hunk 9]
```diff
@@ -113,10 +139,17 @@ private State getState(String type, String value) {
...
         String[] topicElements = getTopicElements(topic);
-        if (topicElements.length != 4)
+        if (topicElements.length < 4)
             throw new IllegalArgumentException("Event creation failed, invalid topic: " + topic);
         return topicElements[2];
     }
+    
+    private String getMemberName(String topic) {
+        String[] topicElements = getTopicElements(topic);
+        if (topicElements.length != 5)
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/events/ItemEventFactory.java]: Shouldn't a check for "lower then" not be sufficient?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java (hunk 6 above)
Comment (chriskn) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: I followed the same approach that is used to notify observers even I don't understand why oberservs are notified if nothing changes. 
See: https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GenericItem.java#L193

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/events/ItemEventFactory.java (hunk 9 above)
Comment (chriskn) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/events/ItemEventFactory.java]: The old implementation checked if the length equals 4. Now, that the topic has one more part, I check if the length equals 5. I thought if the topic has more than 5 parts something is wrong with the topic and a Exception should be thrown because there are no topics which have more than 5 parts

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java (hunk 6 above)
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: * There is an update event that indicates that the state was updated (regardless if the state changed or not).
* There is an changed event that indicates that the state has been changed.

If I want to build a life chart, I want perhaps to see a dot for every state. If the state is not changed, I would like to see a dot because it signals that the end device is still alive and say "yes, it is still 19° C".

If I would react on state changes, I do not want to handle all "updates", I am interested in changes only. I register for changed event and not for update events. If the state changes and there is an update event posted only, but no changed event, there is something broken (to me).

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/events/ItemEventFactory.java (hunk 9 above)
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/events/ItemEventFactory.java]: It is no problem to use "!=" or "<".
But I think at least as the member name is the fifth element, I am also fine with six or more elements.
IMHO the getMemberName does not need to depend on if there has more elements added.
I see that getItemName is using that approach before, but I (personally) would prefer usinng "<".

File: targetplatform/smarthome.target [hunk 10]
```diff
@@ -131,7 +131,7 @@
 </location>
 <location includeAllPlatforms="false" includeConfigurePhase="false" includeMode="planner" includeSource="true" type="InstallableUnit">
 <unit id="javax.jmdns.feature.feature.group" version="3.4.2"/>
-<repository location="http://www.jmdns.org/update-site/3.4.2/"/>
+<repository location="http://www.openhab.org/jmdns/update-site/3.4.2/"/>
```
Comment (kaikreuzer) [inline targetplatform/smarthome.target]: This file should be removed from the commit

**Comment by kaikreuzer** [comment]
There is now a little mismatch between ItemStateChangeEvents and GroupItemStateChangeEvents - the first are sent by the ItemUpdater, while the latter are sent by the GroupItem itself; I think we should have the same mechanism on both.
 So possibly the event logic from ItemUpdater should be moved to GenericItem instead? Would that make sense?

**Comment by kaikreuzer** [comment]
I am also a bit concerned about the GroupItemStateEvents - I expect the event bus to be flooded (and likewise the log files, if these events are logged).
Look e.g. at https://github.com/openhab/openhab2/blob/master/distribution/openhabhome/conf/items/demo.items:
If we receive a state event for `Heating_GF_Living`, this used to be the only event on the bus.
With this new feature now, we would additionally have events on `GF_Living`, `Heating`, `gGF` and `All`.

**Comment by kaikreuzer** [comment]
@maggu2810: What do you think? I fully agree that having GroupItemStateEvents is cleaner, but on the other side, I do not really see a real use case for having all these events - if someone really wants to know, if there are new events, he can also listen on events being sent for the group members.

**Comment by maggu2810** [comment]
I think both aspects are valid.

For encapsulated groups there are a lot of events, it could slow down the performance by factors.
Could we state that the group items are a little bit special (they already are, as this event is missing) and does not send the update event but the change event only?
If the function calculates the avg, min, max, ... using the group item on a diagram will work.
If the encapsulated groups are using all a function that triggers a real state change, we won nothing, but perhaps that is then intention...

On the other hand, we could decide that there are no events for groups at all and you have to watch for the members yourself. But then is such a function (min, max, avg) not really useful.

**Comment by kaikreuzer** [comment]
I would agree that group items are something special and thus we can also have exceptional event handling for them. Having no events is not nice though, since change events will be useful in many situations, like rule triggers, UI updates, etc.

@chriskn Does this sound ok for you? I.e. remove the update events and also adapt the "normal" change events as suggested in https://github.com/eclipse/smarthome/pull/679#issuecomment-165109090?

**Comment by chriskn** [comment]
Sounds ok for me. I will update the PR

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java [hunk 11]
```diff
@@ -243,15 +246,30 @@ public String toString() {
      */
     @Override
     public void stateChanged(Item item, State oldState, State newState) {
-        setState(function.calculate(members));
+        setStateWithouEvent(function.calculate(members));
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: typo: Without

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java [hunk 12]
```diff
@@ -243,15 +246,30 @@ public String toString() {
...
      * @{inheritDoc
      */
     @Override
     public void stateUpdated(Item item, State state) {
-        setState(function.calculate(members));
+        setStateWithouEvent(function.calculate(members));
+        sendGroupStateChangedEvent(item.getName(), this.state, oldGroupState);
+    }
+    
+    private void setStateWithouEvent(State state) {
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: what is the difference of this method to setState()? The look pretty much the same to me...?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java [hunk 13]
```diff
@@ -27,6 +28,8 @@
 
     private final Logger logger = LoggerFactory.getLogger(GroupItem.class);
 
+    private State oldGroupState = UnDefType.NULL;
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: what is this field good for? Why do we have to keep an "old" state and what does "old" exactly mean?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java [hunk 14]
```diff
@@ -243,15 +246,30 @@ public String toString() {
...
+    }
+    
+    private void setStateWithouEvent(State state) {
+        this.oldGroupState = this.state;
+        this.state = state;
+        notifyListeners(oldGroupState, state);
     }
 
+    private void sendGroupStateChangedEvent(String memberName, State newState, State oldState) {
+        if (eventPublisher != null && !oldGroupState.equals(this.state)) {
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: Can't you use the oldState instead of the oldGroupState and newState instead of this.state here? Would be much nicer to refer to the passed parameters than to instance variables.

File: bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/items/events/ItemEventFactoryTest.groovy [hunk 15]
```diff
@@ -107,6 +116,24 @@ class ItemEventFactoryTest extends OSGiTest {
         assertThat itemStateEvent.getSource(), is(SOURCE)
         assertThat itemStateEvent.getItemState(), is(UnDefType.UNDEF)
     }
+    
+    @Test
+    void 'ItemEventFactory creates GroupItemStateChangedEvent UnDefType correctly'() {
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/items/events/ItemEventFactoryTest.groovy]: why UnDefType? Is this a copy&paste remainder?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GenericItem.java [hunk 16]
```diff
@@ -183,20 +183,29 @@ public void setState(State state) {
         State oldState = this.state;
         this.state = state;
         notifyListeners(oldState, state);
+        if (!oldState.equals(state)){
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GenericItem.java]: space missing - is your formatter working?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java [hunk 17]
```diff
@@ -243,7 +243,11 @@ public String toString() {
      */
     @Override
     public void stateChanged(Item item, State oldState, State newState) {
+        State oldGroupState = this.state;
         setState(function.calculate(members));
+        if (!oldGroupState.equals(this.state)) {
+            sendGroupStateChangedEvent(item.getName(), this.state, oldGroupState);
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: hm, didn't we discuss today that this should not send an event as it is in any case already done by stateUpdated()?

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/events/GroupItemStateChangedEvent.java [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+    }
+    
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    public String toString() {
+        return "group " + getItemName() + " changed from " + getOldItemState().toString() + " to "
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/events/GroupItemStateChangedEvent.java]: I prefer short log messages for events. Would suggest to change it to
```
return itemName + " changed from " + oldItemState.toString() + " to " + itemState.toString() + " through " + memberName;
```

File: bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/items/GenericItemTest.groovy [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+ *
+ * @author Christoph Knauf - Initial contribution, event tests 
+ */
+class GenericItemTest {
+
+    List<Event> events = []
+    EventPublisher publisher;
+
+    @Before
+    void setUp(){
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core.test/src/test/groovy/org/eclipse/smarthome/core/items/GenericItemTest.groovy]: space missing - apply formatter

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java (hunk 17 above)
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: If we rely on the fast, that stateUpdated is always called before a stateChanged, why is "setState(function.calculate(members));" called in this function? It is already called in stateUpdated(), too.

Comment (chriskn) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: @kaikreuzer : perhaps I have got that wrong?.
We could leave this method completely unimplemented. It would change nothing on the current event behavior because this.state is equal to oldGoupState (updateState() set it before) and so no event is send.  But IMHO an empty method would lead to  an unexpected behavior for users. In this way the method reacts as expected if we change the observer notification in the future.

Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java]: > But IMHO an empty method would lead to an unexpected behavior for users.

In what sense? An empty method only means that we do not apply any further logic if the member state has changed (and not only updated). And this is the case: For the GroupItem, we really do not care.

**Comment by kaikreuzer** [comment]
Looks all good to me now, thanks!

**Comment by maggu2810** [comment]
This PR reverts my changes on ItemUpdaterOSGiTest.groovy

**Comment by maggu2810** [comment]
It seems not be correctly rebased. What should we do now? The changes done with #755 are now reverted.

**Comment by kaikreuzer** [comment]
Sorry, @maggu2810, I didn't notice. I wonder why it did not show a merge conflict if it wasn't properly rebased...?
@chriskn Could you please have a look and create a follow-up PR that brings back the overwritten changes?

**Comment by maggu2810** [comment]
> I wonder why it did not show a merge conflict if it wasn't properly rebased...?

Just a matter of manual conflict solving or handle rebase / soft reset yourself.
At least there is no conflict (for merging it into upstream), because the changes (that revert the file) are part of this PR.
I have not checked if other files are changed, too.
